src/parser/ast.c

```diff
@@ -291,14 +291,110 @@ AST_Validation _Validate_CREATE_Clause(const AST_QueryExpressionNode* ast, char
+	char undefined_alias[30] = "";
+	
+	if (NULL == ast->setNode) {
+		return AST_VALID;
+	}
+
+	if (!ast->matchNode) {
+		return AST_INVALID;
+	}
+	
+    Vector *update_aliases = NewVector(char*, Vector_Size(ast->setNode->set_elements));
+	for (int i=0; i < Vector_Size(ast->setNode->set_elements); i++) {
+		AST_SetElement *update_entity = NULL;
+        Vector_Get(ast->setNode->set_elements, i, &update_entity);
+		Vector_Push(update_aliases, update_entity->entity->alias);
+	}
+
+	AST_Validation res = _Validate_Aliases_In_Match_Clause(
+		update_aliases, ast->matchNode->graphEntities, undefined_alias);
+	Vector_Free(update_aliases);
+
+	if (res != AST_VALID) {
+		sprintf(*reason, "%s not defined", undefined_alias);
+		return AST_INVALID;
+	}
+
 	return AST_VALID;
 }
 
 AST_Validation _Validate_DELETE_Clause(const AST_QueryExpressionNode* ast, char **reason) {
+	char undefined_alias[30];
+
+	if (NULL == ast->deleteNode) {
+		return AST_VALID;
+	}
+
+	if (!ast->matchNode) {
+		return AST_INVALID;
+	}
+
+	if (_Validate_Aliases_In_Match_Clause(ast->deleteNode->graphEntities, 
+			ast->matchNode->graphEntities, undefined_alias) != AST_VALID) {
+		sprintf(*reason, "%s not defined", undefined_alias);
+		return AST_INVALID;
+	}
+
 	return AST_VALID;
 }
 
 AST_Validation _Validate_RETURN_Clause(const AST_QueryExpressionNode* ast, char **reason) {
+	char undefined_alias[30];
+
+	if (NULL == ast->returnNode) {
+		return AST_VALID;
+	}
+
+	if (!ast->matchNode) {
+		return AST_INVALID;
+	}
+
+	Vector *return_aliases = NewVector(char*, Vector_Size(ast->returnNode->returnElements));
+	for (int i=0; i < Vector_Size(ast->returnNode->returnElements); i++) {
+		AST_ReturnElementNode *return_entity = NULL;
+        Vector_Get(ast->returnNode->returnElements, i, &return_entity);
+		_Get_All_Aliases_From_Expression(return_entity->exp, return_aliases);
+	}
+
+	AST_Validation res = _Validate_Aliases_In_Match_Clause(
+		return_aliases, ast->matchNode->graphEntities, undefined_alias);
+	Vector_Free(return_aliases);
+	
+	if (res != AST_VALID) {
+		sprintf(*reason, "%s not defined", undefined_alias);
+		return AST_INVALID;
+	}
+
+	return AST_VALID;
+}
+
+AST_Validation _Validate_Aliases_In_Match_Clause(const Vector* aliasesToCheck, 
+										         const Vector* matchAliases, 
+										         char* undefined_alias) {
+    /* Check that all the aliases that are in aliasesToCheck Vector exists in the match clause */
+
+	for (int i = 0; i < Vector_Size(aliasesToCheck); i++) {
+        char *alias_to_check = NULL;
+        Vector_Get(aliasesToCheck, i, &alias_to_check);
+
+		int exists = 0;
+		for (int j = 0; j < Vector_Size(matchAliases); j++) {
+			AST_GraphEntity *match_entity = NULL;
+        	Vector_Get(matchAliases, j, &match_entity);
+			if (strcmp(alias_to_check, match_entity->alias) == 0) {
+				exists = 1;
+				break;
+			}
+		}
+
+		if (!exists) {
+			strcpy(undefined_alias, alias_to_check);
```

Comment:
No need to copy,
Simply set undefined_alias to alias_to_check
this is OK because we're performing sprintf in caller.

For that to work undefined_alias should be a char **.
*undefined_alias = alias_to_check.